File: python/crypto_api/elgamal/threshold_elgamal_encryption.py
import math_lib.misc
from crypto_api.elgamal.elgamal_data_classes import ElGamalCipherText, ElGamalKeyParams
from crypto_api.elgamal.elgamal_encryption import ElGamalEncryption
from crypto_api.math_lib.multiplicative_group import MultiplicativeGroup


###TODO: UNDER CONSTRUCTION
class ThresholdElGamalEncryption:

    @staticmethod
    def key_gen(numOfPlayers: int, prime: int, generator: int) -> (int, int, [int]):
        group: MultiplicativeGroup = MultiplicativeGroup(prime)
        polynom_degree: int = numOfPlayers - 1
        x_coords_of_players: [int] = [x for x in range(0, numOfPlayers + 1)]
        delta: [int] = math_lib.misc.calc_lagrange_coeff(x_coords_of_players)
        y_coords_of_players: [int] = [group.random_element() for x in range(0, numOfPlayers)]

        # Index 0 is reserved for the shared secret
        secret_keys: [int] = [0 for x in range(0, numOfPlayers + 1)]
        for n in range(1, len(x_coords_of_players)):
            for p in range(0, numOfPlayers):
                secret_keys[n] += y_coords_of_players[p] * (x_coords_of_players[n] ^ p)

        secret_keys[0]: [int] = ThresholdElGamalEncryption.calc_shared_secret(secret_keys, delta)
        public_key: int = group.pow_mod(generator, secret_keys[0])
        secret_keys[0] = 0

        return delta, public_key, secret_keys

    @staticmethod
    def enc(other_public_key: int, key_params: ElGamalKeyParams, message: int) -> ElGamalCipherText:
        return ElGamalEncryption.enc(other_public_key, key_params, message)

    @staticmethod
    def dec(secret_keys: [int], delta: [int], key_params: ElGamalKeyParams, ciphertext: ElGamalCipherText) -> int:
        shared_secret: int = ThresholdElGamalEncryption.calc_shared_secret(secret_keys, delta)
        return ElGamalEncryption.dec(shared_secret, key_params, ciphertext)

    @staticmethod
    def calc_shared_secret(secret_key: [int], delta: [int]) -> int:
        shared_secret: int = 0
        for i in range(1, len(delta)):
            shared_secret += delta[i] * secret_key[i]
        return shared_secret

File: java/cryptoapi/elgamal/ThresholdElGamalEncryption.java
package cryptoapi.elgamal;

import cryptoapi.math_lib.CyclicMultiplicativeGroup;
import cryptoapi.math_lib.MathMisc;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.stream.Stream;

public class ThresholdElGamalEncryption {

    public static ThresholdElGamalKeyCollection keyGen(int secPar, int numOfPlayers) {
        CyclicMultiplicativeGroup group = new CyclicMultiplicativeGroup(MathMisc.randomBigIntPrime(secPar));
        BigInteger generator = group.getGenerator();
        int polynomDegree = numOfPlayers - 1;

        BigInteger[] xCoordsOfPlayers = new BigInteger[numOfPlayers + 1];
        BigInteger[] delta = MathMisc.calcLagrangeCoeff(xCoordsOfPlayers);
        BigInteger[] yCoordsOfPlayers = new BigInteger[numOfPlayers];

        // Generate random y-coordinates for players
        for (int i = 0; i < numOfPlayers; i++) {
            yCoordsOfPlayers[i] = group.randomElement();
        }

        // Initialize secret keys
        BigInteger[] secretKeys = new BigInteger[numOfPlayers + 1];
        for (int n = 1; n < xCoordsOfPlayers.length; n++) {
            secretKeys[n] = BigInteger.ZERO;
            for (int p = 0; p < numOfPlayers; p++) {
                secretKeys[n] = secretKeys[n].add(yCoordsOfPlayers[p].multiply(xCoordsOfPlayers[n].pow(p)));
            }
        }

        secretKeys[0] = calcSharedSecret(secretKeys, delta);
        BigInteger publicKey = group.powMod(generator, secretKeys[0]);
        secretKeys[0] = BigInteger.ZERO; // Reset the shared secret key

        return new ThresholdElGamalKeyCollection(
                new ElGamalKeyParams(group.m, generator),
                publicKey,
                Arrays.stream(secretKeys, 1, secretKeys.length).toArray(BigInteger[]::new),
                delta
        );
    }

    public static ElGamalCipherText enc(BigInteger otherPublicKey, ElGamalKeyParams keyParams, BigInteger message) {
        return ElGamalEncryption.enc(otherPublicKey, keyParams, message);
    }

    public static BigInteger dec(ThresholdElGamalKeyCollection keyCollection, ElGamalCipherText ciphertext) {
        BigInteger sharedSecret = calcSharedSecret(keyCollection.individualSecretKeys, keyCollection.delta);

        return ElGamalEncryption.dec(
            new ElGamalKeyPair(
                keyCollection.keyParams,
                sharedSecret,
                keyCollection.publicKey
            ),
            ciphertext
        );
    }

    private static BigInteger calcSharedSecret(BigInteger[] secretKeys, BigInteger[] delta) {
        BigInteger sharedSecret = BigInteger.ZERO;
        for (int i = 1; i < delta.length; i++) {
            sharedSecret = sharedSecret.add(delta[i].multiply(secretKeys[i]));
        }
        return sharedSecret;
    }
}

